src/server/journal/streamer.cc

```diff
@@ -11,62 +11,178 @@
+  size_t total_pending = pending_buf_.size() + str.size();
+  if (in_flight_bytes_ == 0 || total_pending > kFlushThreshold) {
+    // because of potential SOO with strings we allocate explicitly on heap
+    uint8_t* buf(new uint8_t[str.size()]);
+
+    // TODO: it is possible to remove these redundant copies if we adjust high level
+    // interfaces to pass reference-counted buffers.
+    memcpy(buf, str.data(), str.size());
+    in_flight_bytes_ += total_pending;
+    io::Bytes src(buf, str.size());
+
+    if (pending_buf_.empty()) {
+      dest_->AsyncWrite(src, [buf, this, len = str.size()](std::error_code ec) {
+        delete[] buf;
+        OnCompletion(ec, len);
+      });
+    } else {
+      iovec v[2];
+      v[0] = IoVec(pending_buf_);
+      v[1] = IoVec(src);
+
+      dest_->AsyncWrite(
+          v, 2,
+          [buf0 = std::move(pending_buf_), buf, this, len = total_pending](std::error_code ec) {
+            delete[] buf;
+            OnCompletion(ec, len);
+          });
+    }
+    return;
   }
+
+  DCHECK_GT(in_flight_bytes_, 0u);
+  DCHECK_LE(pending_buf_.size() + str.size(), kFlushThreshold);
+
+  // Aggregate
+  size_t tail = pending_buf_.size();
+  pending_buf_.resize(pending_buf_.size() + str.size());
+  memcpy(pending_buf_.data() + tail, str.data(), str.size());
 }
 
-void JournalStreamer::WriterFb(io::Sink* dest) {
-  if (auto ec = ConsumeIntoSink(dest); ec) {
+void JournalStreamer::OnCompletion(std::error_code ec, size_t len) {
+  DCHECK_GE(in_flight_bytes_, len);
+
+  DVLOG(2) << "Completing from " << in_flight_bytes_ << " to " << in_flight_bytes_ - len;
+  in_flight_bytes_ -= len;
+  if (ec && !is_stopped_) {
+    is_stopped_ = true;
     cntx_->ReportError(ec);
+  } else if (in_flight_bytes_ == 0 && !pending_buf_.empty() && !is_stopped_) {
+    // If everything was sent but we have a pending buf, flush it.
+    io::Bytes src(pending_buf_);
+    in_flight_bytes_ += src.size();
+    dest_->AsyncWrite(src, [buf = std::move(pending_buf_), this](std::error_code ec) {
+      OnCompletion(ec, buf.size());
+    });
+  }
+
+  // notify ThrottleIfNeeded or WaitForInflightToComplete that waits
+  // for all the completions to finish.
+  // ThrottleIfNeeded can run from multiple fibers in the journal thread.
+  // For example, from Heartbeat calling TriggerJournalWriteToSink to flush potential
+  // expiration deletions and there are other cases as well.
+  waker_.notifyAll();
+}
+
+void JournalStreamer::ThrottleIfNeeded() {
+  if (IsStopped() || !IsStalled())
+    return;
+
+  auto next = chrono::steady_clock::now() + kSinkThrottleTime;
+  std::cv_status status =
+      waker_.await_until([this]() { return !IsStalled() || IsStopped(); }, next);
+  if (status == std::cv_status::timeout) {
+    LOG(WARNING) << "Stream timed out, inflight bytes " << in_flight_bytes_;
+    is_stopped_ = true;
+    cntx_->ReportError(make_error_code(errc::stream_timeout));
   }
```

Comment:
so a replica lag of 0.5 second is "impossible"? There won't be any way to disable this later 🙂

Reply:
I will make it a flag. It's not just a lag. It means we have `in_flight_bytes_` more than a threshold AND we wait on it to decrease for more than 0.5s. But I agree that if we say serialize 4GB value, we won't update in_flight_bytes_ until the whole request has not been sent and it may take quite a while